core/errors/wrappedErrors.go

```diff
@@ -0,0 +1,118 @@
+package errors
+
+import (
+	"errors"
+	"fmt"
+	"runtime"
+)
+
+const skipLevels = 3
+const nilPlaceholder = "<nil>"
+
+type errorWithLocation struct {
+	err      error
+	location string
+}
+
+type wrappableError struct {
+	errsWithLocation []errorWithLocation
+}
+
+// WrapError constructs a WrappableError from an error
+func WrapError(err error) WrappableError {
+	errAsWrappable, ok := err.(WrappableError)
+	if ok {
+		return errAsWrappable
+	}
+	return &wrappableError{
+		errsWithLocation: []errorWithLocation{createErrorWithLocation(err, 2)},
+	}
+}
+
+// WrapWithMessage wraps the target error with a new one, created using the input message
+func (werr *wrappableError) WrapWithMessage(errMessage string) WrappableError {
+	return werr.wrapWithErrorWithSkipLevels(errors.New(errMessage), skipLevels)
+}
+
+// WrapWithStackTrace wraps the target error with a new one, without any message only a stack frame trace
+func (werr *wrappableError) WrapWithStackTrace() WrappableError {
+	return werr.wrapWithErrorWithSkipLevels(errors.New(""), skipLevels)
+}
+
+// WrapWithError wraps the target error with the provided one
+func (werr *wrappableError) WrapWithError(err error) WrappableError {
+	return werr.wrapWithErrorWithSkipLevels(err, skipLevels)
+}
+
+// GetBaseError gets the core error
+func (werr *wrappableError) GetBaseError() error {
+	errorSlice := werr.errsWithLocation
+	return errorSlice[0].err
```

Comment:
not sure if accessing slice without checking it's length is the best approach, but can remaing like this. `errsWithLocation` cannot be an empty slice

Reply:
there is no way we can have a 0 length slice according to the code.